gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagTaskStream.java

```diff
@@ -0,0 +1,255 @@
+import com.typesafe.config.ConfigException;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.gobblin.annotation.Alpha;
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.metrics.event.TimingEvent;
+import org.apache.gobblin.runtime.api.DagActionStore;
+import org.apache.gobblin.runtime.api.MultiActiveLeaseArbiter;
+import org.apache.gobblin.service.ExecutionStatus;
+import org.apache.gobblin.service.modules.flowgraph.Dag;
+import org.apache.gobblin.service.modules.orchestration.proc.KillDagProc;
+import org.apache.gobblin.service.modules.orchestration.task.DagTask;
+import org.apache.gobblin.service.modules.orchestration.task.KillDagTask;
+import org.apache.gobblin.service.modules.spec.JobExecutionPlan;
+import org.apache.gobblin.service.monitoring.JobStatus;
+import org.apache.gobblin.service.monitoring.JobStatusRetriever;
+
+import static org.apache.gobblin.service.ExecutionStatus.ORCHESTRATED;
+import static org.apache.gobblin.service.ExecutionStatus.valueOf;
+
+
+/**
+ * Holds a stream of {@link DagTask} that {@link DagProcessingEngine} would pull from to process, as it is ready for more work.
+ * It provides an implementation for {@link DagManagement} that defines the rules for a flow and job.
+ * Implements {@link Iterable} to provide {@link DagTask}s as soon as it's available to {@link DagProcessingEngine}
+ */
+
+@Alpha
+@Slf4j
+@AllArgsConstructor
+public class DagTaskStream implements Iterable<DagTask>, DagManagement {
+
+  @Getter
+  private final BlockingQueue<DagActionStore.DagAction> dagActionQueue = new LinkedBlockingQueue<>();
+  private FlowTriggerHandler flowTriggerHandler;
+  private DagManagementStateStore dagManagementStateStore;
+  private DagManagerMetrics dagManagerMetrics;
+
+
+  @NotNull
+  @Override
+  public Iterator<DagTask> iterator() {
+    return new Iterator<DagTask>() {
+
+      @Override
+      public boolean hasNext() {
+        return true;
+      }
+
+      @Override
+      public DagTask next() {
+
+        DagTask dagTask = null;
+        while(true) {
+          DagActionStore.DagAction dagAction = take();
+          Properties jobProps = getJobProperties(dagAction);
+          try {
+            MultiActiveLeaseArbiter.LeaseAttemptStatus leaseAttemptStatus = flowTriggerHandler.getLeaseOnDagAction(jobProps, dagAction, System.currentTimeMillis());
+            if(leaseAttemptStatus instanceof MultiActiveLeaseArbiter.LeaseObtainedStatus) {
+              dagTask = createDagTask(dagAction,
+                  (MultiActiveLeaseArbiter.LeaseObtainedStatus) leaseAttemptStatus);
+            }
+            if (dagTask != null) {
+              break; // Exit the loop when dagTask is non-null
+            }
+          } catch (IOException e) {
+            //TODO: need to handle exceptions gracefully
+            throw new RuntimeException(e);
+          }
+        }
+        return dagTask;
```

Comment:
```return createDagTask(...)```
would seem to be more natural control flow, than is the `DagTask dagTask = null;`, the `break` and the `return dagTask`